From ac8a9bc2ef301f1f18772af7d871c257a56ae70c Mon Sep 17 00:00:00 2001
From: Bdale Garbee <bdale@gag.com>
Date: Thu, 30 Aug 2018 21:36:18 -0600
Subject: [PATCH] altos: don't build fox1ihu by default

---
 src/Makefile | 1 -
 1 file changed, 1 deletion(-)

diff --git a/src/Makefile b/src/Makefile
index 0dafb5ec..a1ff84c2 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -25,7 +25,6 @@ SDCCDIRS=\
 	telerepeat-v1.0
 
 ARMM3DIRS=\
-	fox1ihu fox1ihu/flash-loader \
 	easymega-v1.0 easymega-v1.0/flash-loader \
 	telemega-v0.1 telemega-v0.1/flash-loader \
 	telemega-v1.0 telemega-v1.0/flash-loader \
-- 
2.19.0.rc1

